fendermint/vm/topdown/src/sync.rs

```diff
@@ -3,139 +3,245 @@
+        // the genesis epoch of the current subnet and its corresponding block hash.
+        if finality.height == 0 {
+            let genesis_epoch = agent.get_genesis_epoch(subnet_id).await?;
+            let block_hash = agent.get_block_hash(genesis_epoch as u64).await?;
+            finality = IPCParentFinality {
+                height: genesis_epoch as u64,
+                block_hash,
+            };
+            tracing::info!(
+                "no previous finality committed, fetched from genesis epoch: {finality:?}"
+            );
+        }
+
+        return Ok(finality);
+    }
+}
+
+/// Start the polling parent syncer in the background
+pub async fn launch_polling_syncer<T: ParentFinalityStateQuery + Send + Sync + 'static>(
+    subnet_id: &SubnetID,
+    query: T,
+    config: Config,
+    view_provider: Arc<Toggle<CachedFinalityProvider>>,
+    agent: Arc<IPCAgentProxy>,
+) -> anyhow::Result<()> {
+    if !view_provider.is_enabled() {
+        return Err(anyhow!("provider not enabled, enable to run syncer"));
+    }
+
+    let query = Arc::new(query);
+
+    let finality = query_starting_finality(subnet_id, &query, &agent).await?;
+
+    atomically(|| view_provider.set_new_finality(finality.clone())).await;
+
+    let poll = PollingParentSyncer::new(subnet_id.clone(), config, view_provider, agent, query);
+    poll.start();
+
+    Ok(())
+}
+
+impl<T> PollingParentSyncer<T> {
     pub fn new(
+        subnet_id: SubnetID,
         config: Config,
-        parent_view_provider: Arc<InMemoryFinalityProvider>,
+        parent_view_provider: Arc<Toggle<CachedFinalityProvider>>,
         agent: Arc<IPCAgentProxy>,
+        query: Arc<T>,
     ) -> Self {
         Self {
+            subnet_id,
             config,
             parent_view_provider,
             agent,
+            committed_state_query: query,
         }
     }
 }
 
-impl PollingParentSyncer {
+impl<T: ParentFinalityStateQuery + Send + Sync + 'static> PollingParentSyncer<T> {
     /// Start the parent finality listener in the background
-    pub fn start(self) -> anyhow::Result<()> {
+    pub fn start(self) {
         let config = self.config;
         let provider = self.parent_view_provider;
         let agent = self.agent;
+        let query = self.committed_state_query;
+        let subnet_id = self.subnet_id;
 
         tokio::spawn(async move {
             loop {
-                // Syncing with parent with the below steps:
-                // 1. Get the latest height in cache or latest height committed increment by 1 as the
-                //    starting height
-                // 2. Get the latest chain head height deduct away N blocks as the ending height
-                // 3. Fetches the data between starting and ending height
-                // 4. Update the data into cache
-                if let Err(e) = sync_with_parent(&config, &agent, &provider).await {
+                let mut interval =
```

Comment:
Any reason this is created inside the loop rather than out?

Reply:
copy paste error...